tensorflow/python/eager/def_function_test.py

```diff
@@ -917,6 +917,50 @@ def f(x):
     self.assertLen(logs.output, 1)
     self.assertIn('Tracing is expensive', logs.output[0])
 
+  def test_type_annotation_basic(self):
+    def f(x: ops.Tensor[dtypes.Int32], y: ops.Tensor[dtypes.Float32]):
+      return x
+    signature = [tensor_spec.TensorSpec(None, dtypes.int32),
+                 tensor_spec.TensorSpec(None, dtypes.float32)]
+    def_function.function(f, input_signature=signature)
+
+  def test_mismatch_type_annotation(self):
+    msg = 'Type annotation does not match input_signature'
+    with self.assertRaisesRegexp(ValueError, msg):
+      def f(x: ops.Tensor[dtypes.Int32], y: ops.Tensor[dtypes.String]):
+        return x
+      signature = [tensor_spec.TensorSpec(None, dtypes.int32),
+                   tensor_spec.TensorSpec(None, dtypes.float32)]
+      def_function.function(f, input_signature=signature)
+
+  def test_non_tensor_type_annotation(self):
+    def f(x: int):
+      return x
+    signature = [tensor_spec.TensorSpec(None, dtypes.int32)]
+    def_function.function(f, input_signature=signature)
+
+    msg = 'Type annotation does not match input_signature'
```

Comment:
Is this also a "non-tensor annotation"?

Reply:
The annotation is a Tensor, but the parameter within the Tensor annotation is incorrect. With a more informative error message, the error raised here will be clearer.